### PR TITLE
dry-run: Create Options with dryRun for POST/PUT/PATCH

### PR DESCRIPTION
**Pull request by apelisse**
**What this PR does / why we need it**:
Create new options for Create and Update (through POST/PUT/PATCH).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/registry/apps/deployment/registry.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func (s *storage) GetDeployment(ctx context.Context, deploymentID string, option
 }
 
 func (s *storage) CreateDeployment(ctx context.Context, deployment *extensions.Deployment, createValidation rest.ValidateObjectFunc) (*extensions.Deployment, error) {
-	obj, err := s.Create(ctx, deployment, createValidation, false)
+	obj, err := s.Create(ctx, deployment, createValidation, &metav1.CreateOptions{})
```
Comment (lavalamp) [inline pkg/registry/apps/deployment/registry.go]: Maybe make a metav1.CreateOptionsTODO() function so you can be sure to find and fix all of these after the fact? If you go this route, anyway.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -453,6 +453,36 @@ type DeleteOptions struct {
...
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CreateOptions may be provided when creating an API object.
+type CreateOptions struct {
+	TypeMeta `json:",inline"`
+
+	// DryRun decides if the request should be completed or not.
+	// +optional
+	DryRun bool `json:"dryRun,omitempty" protobuf:"varint,1,opt,name=dryRun"`
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: I think the decision from the call yesterday was to make this an enumerated string.

**Comment by deads2k** [comment]
@sttts I'm behind at the moment.  I've got a [comment](https://github.com/kubernetes/kubernetes/pull/65154#discussion_r197108505) on the prereq pull and the API here needs an update.  Got time while I deal with the downstream?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 2 above)
Comment (apelisse) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Absolutely, I haven't taken the time to update yet.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 3]
```diff
@@ -468,6 +468,10 @@ type CreateOptions struct {
 	// DryRun decides if the request should be completed or not.
 	// +optional
 	DryRun bool `json:"dryRun,omitempty" protobuf:"varint,1,opt,name=dryRun"`
+
+	// If IncludeUninitialized is specified, the object may be
+	// returned without completing initialization.
+	IncludeUninitialized bool `json:"includeUninitialized,omitempty" protobuf:"varint,2,opt,name=includeUninitialized"`
```
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: this surprises me. How is this related to dry-run? Why do we expose that via an API "suddenly"?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go [hunk 4]
```diff
@@ -260,6 +260,14 @@ func DeleteCollection(r rest.CollectionDeleter, checkBody bool, scope RequestSco
 
 				ae := request.AuditEventFrom(ctx)
 				audit.LogRequestObject(ae, obj, scope.Resource, scope.Subresource, scope.Serializer)
+			} else {
+				if values := req.URL.Query(); len(values) > 0 {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go]: this is guarded by `if checkBody`. Looks odd.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go]: In the delete handler we call the guarding var `allowsOptions`.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Also note that initializers go away (if I have understood @smarterclayton correctly). We should not expose a parameter for something that is planned to be removed.

Comment (sttts) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Ok, found it: https://github.com/kubernetes/kubernetes/pull/65105/commits/69e757ae6077d31376512e8b6b3e54fbe90a30f6#diff-7861f33474e4eb8d21cf1d33de5e4d4eL114 had the logic before, with a manually parsed get var. So nvm my comment above.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 5]
```diff
@@ -111,7 +111,9 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
 		}
 
 		// TODO: replace with content type negotiation?
-		includeUninitialized := req.URL.Query().Get("includeUninitialized") == "1"
+		options := &metav1.CreateOptions{
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: Shouldn't this be using the parameter decoder? This looks hacky.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 6]
```diff
@@ -80,6 +81,24 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
...
+		if values := req.URL.Query(); len(values) > 0 {
+			// TODO: replace with content type negotiation?
+			// "1" is accepted as "true" for backward compatibility.
+			for i, v := range values["includeUninitialized"] {
+				if v == "1" {
+					values["includeUninitialized"][i] = "true"
+				}
+			}
+			// Timeout is not part of the options, drop it.
+			delete(values, "timeout")
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: DecodeParameters ignores params that are not in the struct (necessary for new clients speaking to old servers). So, no need for the delete.

File: pkg/master/reconcilers/lease.go [hunk 7]
```diff
@@ -221,7 +221,8 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
 	}
 
 	glog.Warningf("Resetting endpoints for master service %q to %v", serviceName, masterIPs)
-	return r.endpointRegistry.UpdateEndpoints(ctx, e, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc)
+	// PLEASE_PAY_ATTENTION(apelisse): No idea if this is what we need to do here. Does it make sense to have DryRun here anyway?
```
Comment (sttts) [inline pkg/master/reconcilers/lease.go]: this is the leader election code which speaks directly to the registry to update the endpoint. So this certainly does not need do a dry-run here. It is very very special purpose.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 8]
```diff
@@ -89,6 +90,17 @@ func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface
 			return
 		}
 
+		options := &metav1.UpdateOptions{}
+		if values := req.URL.Query(); len(values) > 0 {
+			// Timeout is not part of the options, drop it.
+			delete(values, "timeout")
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: same here, DecodeParameters drops unknown fields.

**Comment by sttts** [comment]
Looks generally good. Comments for @jennybuckley's change in her PR.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go (hunk 4 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/delete.go]: Yeah, I'm not sure what the point of these variable is, but it looks like "checkBody" and  "allowOptions" have the same goal here?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go (hunk 8 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: Fixed

File: pkg/master/reconcilers/lease.go (hunk 7 above)
Comment (apelisse) [reply pkg/master/reconcilers/lease.go]: Fixed

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go (hunk 6 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: Fixed

**Comment by apelisse** [comment]
@sttts Feel free to take another look. Still needs to be rebased, but I'm waiting for Jenny's PR to be rebased first.

I've added a couple of clean-up commits that are somewhat unrelated. I've also addressed your comment and changed the dryRun flag to a string.
Thanks

File: pkg/master/reconcilers/lease.go [hunk 9]
```diff
@@ -221,7 +221,8 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
 	}
 
 	glog.Warningf("Resetting endpoints for master service %q to %v", serviceName, masterIPs)
-	return r.endpointRegistry.UpdateEndpoints(ctx, e, rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc)
+	// Use default options, dry-run doesn't make sense here.
```
Comment (sttts) [inline pkg/master/reconcilers/lease.go]: would even drop the dry-run part. Dry-run=false is the default case. And it's obvious that we want to do a real update there. The sentence confuses more than it helps.

File: pkg/registry/apps/deployment/storage/storage_test.go [hunk 10]
```diff
@@ -296,7 +296,7 @@ func TestStatusUpdate(t *testing.T) {
 		},
 	}
 
-	if _, _, err := storage.Status.Update(ctx, update.Name, rest.DefaultUpdatedObjectInfo(&update), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc); err != nil {
+	if _, _, err := storage.Status.Update(ctx, update.Name, rest.DefaultUpdatedObjectInfo(&update), rest.DefaultUpdateConfig(rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, false), &metav1.UpdateOptions{}); err != nil {
```
Comment (sttts) [inline pkg/registry/apps/deployment/storage/storage_test.go]: @jennybuckley I haven't taken another look on your PR, yet. From the changes here: wouldn't a constant like `UpdateNoDryRun` (and a `type UpdateMode bool` type) be much easier to read? This `false` or even a `true` value looks far too innocent.

**Comment by sttts** [comment]
@apelisse can you squash your commits? It's a bit hard to review with @jennybuckley's update signature changes all over the place (and Github's inability to show a commit range only).

**Comment by apelisse** [comment]
Squashed. Except for the code generation, that I've left in a separate commit.

**Comment by apelisse** [comment]
@sttts Rebased and ready for merge!

**Comment by sttts** [comment]
/assign @liggitt 

for the registry changes.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 11]
```diff
@@ -80,6 +81,21 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
 			return
 		}
 
+		options := &metav1.CreateOptions{}
+		values := req.URL.Query()
+		// TODO: replace with content type negotiation?
+		// "1" is accepted as "true" for backward compatibility.
+		for i, v := range values["includeUninitialized"] {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: this could be moved into a helper maybe. Also, do we regression test this?

File: pkg/api/testing/serialization_test.go [hunk 12]
```diff
@@ -156,8 +156,10 @@ var nonRoundTrippableTypes = sets.NewString(
...
 	// ListOptions is now part of the meta group
 	"ListOptions",
-	// Delete options is only read in metav1
+	// DeleteOptions, CreateOptions and UpdateOptions are only read in metav1
 	"DeleteOptions",
+	"CreateOptions",
+	"UpdateOptions",
 )
 
 var commonKinds = []string{"Status", "ListOptions", "DeleteOptions", "ExportOptions"}
```
Comment (liggitt) [inline pkg/api/testing/serialization_test.go]: do CreateOptions/UpdateOptions need to be added to this list?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 13]
```diff
@@ -418,6 +418,13 @@ const (
 	DeletePropagationForeground DeletionPropagation = "Foreground"
 )
 
+// Possible values for DryRun flag.
+const (
+	// DryRunAll means that the entire chain should be verified: admission
+	// chain, validation, merging, everything.
+	DryRunAll = "All"
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: if this is what we call this, let's make sure we agree that any new stages in the api pipeline in the future will get included in this mode, so that we never ever ever have to have a repeat of the "all does not mean all" issue. cc @smarterclayton

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 14]
```diff
@@ -453,6 +460,36 @@ type DeleteOptions struct {
 	// foreground.
 	// +optional
 	PropagationPolicy *DeletionPropagation `json:"propagationPolicy,omitempty" protobuf:"varint,4,opt,name=propagationPolicy"`
+
+	// DryRun decides if the request should be completed or not.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: make the documentation match the serialized field name: `dryRun`

the documentation for this field should:
* indicate the guarantee of setting the field (the mutation will not be persisted)
* indicate what the recognized values mean ("All" means ____)
* indicate what will be done with unrecognized values

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go (hunk 11 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: Convert_Slice_string_To_bool should already handle this conversion

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go [hunk 15]
```diff
@@ -106,10 +115,30 @@ func PatchResource(r rest.Patcher, scope RequestScope, admit admission.Interface
 		)
 
 		userInfo, _ := request.UserFrom(ctx)
-		staticAdmissionAttributes := admission.NewAttributesRecord(nil, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Update, userInfo)
+		staticAdmissionAttributes := admission.NewAttributesRecord(
```
[the remark this answers is not in the document]
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go]: No, it hasn't. Sorry, I initially had multiple commits, and I have a couple of pure cosmetic changes in the middle that got squashed together.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 13 above)
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: I don't think this is referenced anywhere yet. If we want to drop this from this PR to get the signature/plumbing changes in, and continue discussion of the semantics of the parameter, that would be fine with me

File: pkg/api/testing/serialization_test.go (hunk 12 above)
Comment (apelisse) [reply pkg/api/testing/serialization_test.go]: I don't know, but let's do it for consistency.

**Comment by apelisse** [comment]
@liggitt Updated!

File: pkg/api/testing/serialization_test.go (hunk 12 above)
Comment (sttts) [reply pkg/api/testing/serialization_test.go]: GetOptions

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go [hunk 16]
```diff
@@ -110,17 +120,14 @@ func createHandler(r rest.NamedCreater, scope RequestScope, admit admission.Inte
 			}
 		}
 
-		// TODO: replace with content type negotiation?
-		includeUninitialized := req.URL.Query().Get("includeUninitialized") == "1"
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/create.go]: can you verify this got populated/converted properly ("1" and "true" set to true, unset or anything else set to false)?

**Comment by liggitt** [comment]
/hold

forgot we didn't settle the param type... string vs []string

**Comment by deads2k** [comment]
> forgot we didn't settle the param type... string vs []string

Heh, I just thought to ask about that too.

/approve cancel

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 17]
```diff
@@ -453,6 +453,42 @@ type DeleteOptions struct {
 	// foreground.
 	// +optional
 	PropagationPolicy *DeletionPropagation `json:"propagationPolicy,omitempty" protobuf:"varint,4,opt,name=propagationPolicy"`
+
+	// Whether the request should be completed and modifications will not be persisted.
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: I'd suggest something more like this:

> When present, indicates that modifications should not be persisted.
> An invalid or unrecognized dryRun directive will result in a BadRequest response and no further processing of the request.

and then we can add details about what directives are supported as we determine that

**Comment by deads2k** [comment]
type looks correct now.

/approve

**Comment by roycaihw** [comment]
It looks like we added `meta.v1.CreateOptions` and `meta.v1.UpdateOptions` in this PR, but they weren't reflected in our openapi spec. The `dryRun` query parameter was only added to the delete operations in the spec. I'm wondering if this was intended?
